File: src/SpaceGame/objects/components.py
'''
Created on 2013-11-03

@author: Nich
'''

from model.base import Base
from sqlalchemy import Column, String, Integer, Float, ForeignKey, PickleType, Boolean
from sqlalchemy.orm import relationship, backref
import time


class NetworkMessages(object):

    def __init__(self, entity_id):
        self.entity_id = entity_id
        self.msg = []

    def send(self, text):
        import Systems.NetworkMessageSystem as NetMes
        message = NetMes.NetworkMessage(self.entity_id, text)
        self.msg.append(message)


class PlayerControlled(Base):
    __compname__ = "player_controlled"
    __tablename__ = "player_controlled"
    id = Column(Integer, primary_key=True)

    entity_id = Column(String, ForeignKey("entity.id"))
    pid = Column(String)

    def __init__(self, entity_id, pid=0):
        self.entity_id = entity_id
        self.pid = pid


class PlayerInput():

    def __init__(self, entity_id, data=None):
        self.entity_id = entity_id
        self.data = [] if data is None else data


class DBPosition(Base):
    __compname__ = "db_position"
    __tablename__ = "db_position"
    id = Column(Integer, primary_key=True)
    entity_id = Column(String, ForeignKey("entity.id"))
    x = Column(Float)
    y = Column(Float)

    def __init__(self, entity_id, x, y):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class Persisted():

    def __init__(entity_id, persisted_properties=None):
        self.entity_id = entity_id
        # PersistedProperties is a dict mapping component name to the name of
        # the db_component that persists it
        self.persisted_properties = {} if persisted_properties == None else persisted_properties


class InstanceComponents(Base):
    __compname__ = "instance_components"
    __tablename__ = "instance_components"

    id = Column(Integer, primary_key=True)
    entity_id = Column(String, ForeignKey("entity.id"))
    components = Column(String)

    def __init__(self, entity_id, components="{}"):
        self.entity_id = entity_id
        self.components = components


class Position():

    def __init__(self, entity_id, x=0, y=0):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class Area():

    def __init__(self, entity_id, radius):
        self.entity_id = entity_id
        self.radius = radius


class Collidable():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Pickup():

    def __init__(self, entity_id, item_id, qty):
        self.entity_id = entity_id
        self.item_id = item_id
        self.qty = qty


class Colliding():

    def __init__(self, entity_id, collisions):
        self.entity_id = entity_id
        self.collisions = [] if not collisions else collisions


class Sector():
    '''
    A sector is position / 2500
    neightbours has the list of all entities in the sector
    '''

    def __init__(self, entity_id, sx, sy, fx, fy, neighbours, fine_neighbours):
        self.entity_id = entity_id
        self.sx = sx
        self.sy = sy
        self.fx = fx
        self.fy = fy
        self.neighbours = neighbours
        self.fine_neighbours = fine_neighbours


class Shop():

    def __init__(self, entity_id, shop_data):
        self.entity_id = entity_id
        self.shop_data = shop_data


class ShopSpec():

    def __init__(self, entity_id, shop_data):
        self.entity_id = entity_id
        self.shop_data = shop_data


class Money():

    def __init__(self, entity_id, money):
        self.entity_id = entity_id
        self.money = money


class Inventory():

    def __init__(self, entity_id, inventory):
        self.entity_id = entity_id
        self.inv = inventory


class InventoryMass():

    def __init__(self, entity_id, inventory_mass=0):
        self.entity_id = entity_id
        self.inventory_mass = inventory_mass


class Transaction():

    def __init__(self, entity_id, transactions):
        self.entity_id = entity_id
        self.transactions = transactions


class Velocity():

    def __init__(self, entity_id, x=0, y=0):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class Acceleration():

    def __init__(self, entity_id, x=0, y=0):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class Force():

    def __init__(self, entity_id, x=0, y=0):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class Rotation():

    def __init__(self, entity_id, rotation=0):
        self.entity_id = entity_id
        self.rotation = rotation


class Mass():

    def __init__(self, entity_id, mass=200):
        self.entity_id = entity_id
        self.mass = mass


class Type():

    def __init__(self, entity_id, type):
        self.entity_id = entity_id
        self.type = type


class PhysicsUpdate():

    def __init__(self, entity_id, last_update=None):
        self.entity_id = entity_id
        import time
        self.last_update = time.time() * 1000 if last_update is None else last_update
        self.packets = []


class GameStateRequest():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Shooting():

    def __init__(self, entity_id, firing_rate, inputs=None):
        self.entity_id = entity_id
        self.inputs = inputs if not inputs is None else []
        self.firing_rate = firing_rate


class ShootingVars():

    def __init__(self, entity_id):
        self.entity_id = entity_id
        self.last_update = time.time() * 1000
        self.bullets_fired = 0
        # Make none to use firing rate as a default since 0 is valid
        self.residual_cooldown = None


class Mining():

    def __init__(self, entity_id, time):
        self.entity_id = entity_id
        self.time = time


class Minable():

    def __init__(self, entity_id, products):
        self.entity_id = entity_id
        self.products = products


class Proximity():

    def __init__(self, entity_id, proximity_map=None):
        self.entity_id = entity_id
        self.proximity_map = {} if proximity_map is None else proximity_map


class ServerUpdated():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class StateHistory():

    def __init__(self, entity_id):
        self.entity_id = entity_id
        self.history = []


class Camera():

    def __init__(self, entity_id, radius):
        self.new_entities = {}
        self.tracked_entites = {}
        self.leaving_entities = {}
        self.radius = radius
        self.entity_id = entity_id


class Renderable():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Processor():

    def __init__(self, entity_id, processes, last_update=None):
        self.entity_id = entity_id
        self.processes = processes
        self.last_update = time.time() * 1000 if last_update is None else last_update


class Expires():

    def __init__(self, entity_id, expiry_time_ms, creation_time):
        self.entity_id = entity_id
        self.expiry_time_ms = expiry_time_ms
        self.creation_time = creation_time


class Target():

    def __init__(self, entity_id, target_id):
        self.entity_id = entity_id
        self.id = target_id


class OrientTowardsTarget():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class ShootAtTarget():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class ProximityTargetBehaviour():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class PlayerProximityTargetBehaviour():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Health():

    def __init__(self, entity_id, health, max_health, base_health=None):
        self.entity_id = entity_id
        self.health = health
        self.max_health = max_health
        self.base_health = max_health if base_health is None else base_health


class CollisionDamage():

    def __init__(self, entity_id, damage):
        self.entity_id = entity_id
        self.damage = damage


class CollisionVelocityDamage():

    def __init__(self, entity_id, damage, min_velocity=0, max_velocity=100):
        self.entity_id = entity_id
        self.damage = damage
        self.min_velocity = min_velocity
        self.max_velocity = max_velocity


class CollisionMovement():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class ClientSync():

    def __init__(self, entity_id, sync_key):
        self.entity_id = entity_id
        self.sync_key = sync_key


class Animated():

    def __init__(self, entity_id, update_rate):
        self.entity_id = entity_id
        self.update_rate = update_rate


class Event():

    def __init__(self, entity_id, script, cooldown, initial_cooldown=0):
        self.entity_id = entity_id
        self.script = script
        self.cooldown = initial_cooldown
        self.cooldown_time = cooldown


class EventActive():

    def __init__(self, entity_id):
        self.entity_id = entity_id
        self.triggerer = None


class EventProximityTrigger():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class NoSync():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Moved():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class Impulses():

    def __init__(self, entity_id):
        self.entity_id = entity_id
        self.left = 0
        self.right = 0
        self.shoot = 0
        self.thrust = 0
        self.mine = 0
        self.brake = 0


class Allies():

    def __init__(self, entity_id, team):
        self.entity_id = entity_id
        self.team = team


class Home():

    def __init__(self, entity_id, x, y):
        self.entity_id = entity_id
        self.x = x
        self.y = y


class AIReturnHome():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class AvoidShootingAllies():

    def __init__(self, entity_id):
        self.entity_id = entity_id


class AppliedUpgrades():
    # Upgrades are stored as key:data, where the key looks up the function and
    # the function is called with node, data

    def __init__(self, entity_id, upgrades=None):
        self.entity_id = entity_id
        self.upgrades = {} if upgrades is None else upgrades


class ApplyUpgrade():

    def __init__(self, entity_id, upgrade_name, upgrade_id):
        self.entity_id = entity_id
        self.upgrade_id = upgrade_id
        self.upgrade_name = upgrade_name


class Weapon():

    def __init__(self, entity_id, type):
        self.entity_id = entity_id
        self.type = type


class Dead():

    def __init__(self, entity_id):
        self.entity_id = entity_id


components = {
    "acceleration": Acceleration,
    "ai_return_home": AIReturnHome,
    "allies": Allies,
    "animated": Animated,
    "applied_upgrades": AppliedUpgrades,
    "apply_upgrade": ApplyUpgrade,
    "area": Area,
    "avoid_shooting_allies": AvoidShootingAllies,
    "camera": Camera,
    "client_sync": ClientSync,
    "collidable": Collidable,
    "colliding": Colliding,
    "collision_damage": CollisionDamage,
    "collision_movement": CollisionMovement,
    "collision_velocity_damage": CollisionVelocityDamage,
    "dead": Dead,
    "event": Event,
    "event_active": EventActive,
    "event_proximity_trigger": EventProximityTrigger,
    "expires": Expires,
    "force": Force,
    "game_state_request": GameStateRequest,
    "health": Health,
    "home": Home,
    "impulses": Impulses,
    "inventory": Inventory,
    "inventory_mass": InventoryMass,
    "mass": Mass,
    "minable": Minable,
    "mining": Mining,
    "money": Money,
    "moved": Moved,
    "network_messages": NetworkMessages,
    "no_sync": NoSync,
    "orient_towards_target": OrientTowardsTarget,
    "persisted": Persisted,
    "physics_update": PhysicsUpdate,
    "pickup": Pickup,
    "player_controlled": PlayerControlled,
    "player_input": PlayerInput,
    "player_proximity_target_behaviour": PlayerProximityTargetBehaviour,
    "position": Position,
    "processor": Processor,
    "proximity": Proximity,
    "proximity_target_behaviour": ProximityTargetBehaviour,
    "renderable": Renderable,
    "rotation": Rotation,
    "sector": Sector,
    "server_updated": ServerUpdated,
    "shoot_at_target": ShootAtTarget,
    "shooting": Shooting,
    "shooting_vars": ShootingVars,
    "shop": Shop,
    "shop_spec": ShopSpec,
    "state_history": StateHistory,
    "target": Target,
    "transaction": Transaction,
    "type": Type,
    "velocity": Velocity,
    "weapon": Weapon
}

db_components = {
    "instance_components": InstanceComponents,
    "db_position": DBPosition
}

File: src/SpaceGame/static/js/components.js
var components = (function(){ 
	var Position = function(entity_id, data) {
		this.entity_id = entity_id
		this.x = data.x
		this.y = data.y
	};
	var Velocity = function(entity_id, data) {
		this.entity_id = entity_id
		this.x = data.x
		this.y = data.y
	};
	var Acceleration = function(entity_id, data) {
		this.entity_id = entity_id,
		this.x = data.x,
		this.y = data.y
	};
	var Force = function(entity_id, data) {
		this.entity_id = entity_id
		this.x = data.x
		this.y = data.y
	};
	var Thrust = function(entity_id, data) {
		this.entity_id = entity_id;
		this.thrust = data.thrust;
	};
	var Mass = function(entity_id, data) {
		this.entity_id = entity_id;
		this.mass = data.mass;
	};

	var Rotation = function(entity_id, data) {
		this.entity_id = entity_id;
		this.rotation = data.rotation;
	};

	var Renderable = function(entity_id, data) {
		this.entity_id = entity_id;
		this.spritesheet = data.spritesheet;
		this.image = data.image;
		this.width = data.width;
		this.height = data.height;
		//this.transform = data.transform
	};
	var State = function(entity_id, data) {
		this.entity_id = entity_id;
		this.state = data.state;
	};
	var Camera = function(entity_id) {
		this.entity_id = entity_id;
	};

	var ShipControl = function(entity_id, data) {
		this.entity_id = entity_id;
		this.left = data.left;
		this.right = data.right;
		this.thrust = data.thrust;
		this.brake = data.brake;
		this.dt = data.dt;
	}

	var Control = function(entity_id, data) {
		this.entity_id = entity_id;
		this.left = data.left;
		this.right = data.right;
		this.thrust = data.thrust;
		this.brake = data.brake;
		this.shoot = data.shoot;
		this.dt = data.dt;
		this.time = data.time;
	};
	var PlayerControlled = function(entity_id) {
		this.entity_id = entity_id;
	}

	var ServerControlled = function(entity_id) {
		this.entity_id = entity_id;
	}

	var ServerUpdate= function(entity_id, data) {
		this.entity_id = entity_id;
		this.data = data.data;
	}
	
	var Inputs = function(entity_id, data) {
		this.entity_id = entity_id;
		this.inputs = data.inputs;
	}

	var Area = function(entity_id, data) {
		this.entity_id = entity_id;
		this.radius = data.radius;
	}

	var Animated = function(entity_id, data) {
		this.entity_id = entity_id;
		this.update_rate = data.update_rate;
		this.frame = data.frame ? data.frame:0;
		this.residual_cooldown = 0
	}

	var ToBeRemoved = function(entity_id) {
		this.entity_id = entity_id;
	}

	var Health = function(entity_id, data) {
		this.entity_id = entity_id;
		this.health = data.health;
		this.max_health = data.max_health;
	}

	var Type = function(entity_id, data) {
		this.entity_id = entity_id;
		this.type = data.type;
	}

	var Shooting = function(entity_id, data) {
		this.entity_id = entity_id;
		this.input = data.input;
		this.firing_rate = data.firing_rate;
	
	}

	var ShootingVars = function(entity_id, data) {
		this.entity_id = entity_id;
		this.last_update = Date.now();
		this.bullets_fired = 0;
		this.residual_cooldown = null; //Make null so it can be init to variable firing rate
	}

	var Expires = function(entity_id, data) {
		this.entity_id = entity_id
		this.expiry_time_ms = data.expiry_time_ms;
		this.creation_time = data.creation_time;
	}

	var ServerSync = function(entity_id, data) {
		this.entity_id = entity_id
		this.sync_key = data.sync_key
	}

	var ClientSync = function(entity_id, data) {
		this.entity_id = entity_id
		this.sync_key = data.sync_key
	}

	var PlayerCreated = function(entity_id, data) {
		this.entity_id = entity_id
	}

	var Colliding = function(entity_id, data) {
		this.entity_id = entity_id
		this.collisions = []
	}

	var Collidable = function(entity_id, data) {
		this.entity_id = entity_id
	}

	var Mining = function(entity_id, data) {
		this.entity_id = entity_id;
	}

	var Minable = function(entity_id, data) {
		this.entity_id = entity_id;
	}

	var components = {
		"position":Position,
		"velocity":Velocity,
		"acceleration":Acceleration,
		"force":Force,
		"thrust":Thrust,
		"mass": Mass,
		"rotation": Rotation,
		"renderable": Renderable,
		"camera": Camera,
		"control": Control, 
		"player": PlayerControlled,
		"server_controlled": ServerControlled,
		"server_update": ServerUpdate,
		"inputs":Inputs,
		"area": Area,
		"animated":Animated,
		"to_be_removed": ToBeRemoved,
		"health": Health,
		"type": Type,
		"ship_control": ShipControl,
		"shooting": Shooting,
		"shooting_vars": ShootingVars,
		"expires": Expires,
		"server_sync": ServerSync,
		"client_sync": ClientSync,
		"player_created": PlayerCreated,
		"collidable": Collidable,
		"colliding": Colliding,
		"minable": Minable,
		"mining": Mining
	};

	return components
})();

var EntityManager = (function(){
	function EntityManager() {

		return this;
	}

	EntityManager.prototype.init = function() {
		this.last_id = 0;
	}

	EntityManager.prototype.new_entity = function() {
		return ++this.last_id;
	}
 
	return EntityManager;
})();

var NodeFactory = (function(entity_manager, component_manager) {
	function NodeFactory(entity_manager, component_manager) {
		this.entity_manager = entity_manager
		this.component_manager = component_manager;
	}

	NodeFactory.prototype.init = function() {

	}

	NodeFactory.prototype.create_node = function(components, entity_id) {
		let n = Node(this.component_manager);

		e_id = entity_id ? entity_id : this.entity_manager.new_entity();

		n.init(e_id, []);

		let component_ids = Object.keys(components);
		for(let c in component_ids) {
			let comp = component_ids[c];
			n.add_or_attach(comp, components[comp]);
		}
		return n;
	}

	NodeFactory.prototype.create_node_list = function(mandatory, optional, entity_ids) {
		optional = optional ? optional : [];

		let nodes = [];
		entities = this.component_manager.get_entities_with_components(mandatory);

		let all_components = []
		all_components = all_components.concat(mandatory);
		all_components = all_components.concat(optional);

		for(let i = 0; i < entities.length; i++) {
			var n = Node(this.component_manager);
			let entity_id = entities[i];
			n.init(entity_id, []);
			for(let j = 0; j < all_components.length; j++) {
				let comp = all_components[j];
				if(this.component_manager.entity_has_component(entity_id, comp)) {
					n.add_or_attach(comp, {});
				}
			}
			nodes.push(n);

		}
		return nodes;

	}
	var Node = (function(component_manager) {

		function Node(component_manager) {
			this.component_manager = component_manager;
			return this;
		}

		Node.prototype.init = function(entity_id, components) {
			this.id = entity_id
			this.component_list = Object.keys(components)
			for(let i in this.component_list){
				c = this.component_list[i];
				this[c] = components[c];
			}
		}

		Node.prototype.has = function(compo) {
			return this.component_list.indexOf(compo) > -1;
		};

		Node.prototype.entity_has = function(comp) {
			return this.component_manager.entity_has_component(this.id, comp);
		}

		Node.prototype.add_or_attach = function(comp, comp_data) {
			if(this.has(comp)){
				return;
			}
			if(!this.component_manager.entity_has_component(this.id, comp)) {
				this.component_manager.add_component_to_entity(this.id, comp, comp_data)
			}
			this.component_list.push(comp);
			this[comp] = this.component_manager.get_component_data_for_entity(this.id, comp);

		};

		Node.prototype.add_or_update = function(comp, comp_data) {
			this.component_manager.update_component_for_entity(this.id, comp, comp_data);
			this.component_list.push(comp);
			this[comp] = this.component_manager.get_component_data_for_entity(this.id, comp);
		};

		Node.prototype.remove_component = function(comp) {
			delete this[comp];
			this.component_list = this.component_list.filter(function(i){i !== comp});
		};

		Node.prototype.delete_component = function(comp) {
			this.remove_component(comp);
			this.component_manager.remove_component_from_entity(this.id, comp);
		}

		Node.prototype.delete_all_components = function(comp) {
			for(let i = 0; i < this.component_list.length; i++) {
				this.remove_component(this.component_list[i]);
			}
			this.component_manager.remove_all_components_from_entity(this.id);
		}	

		return new Node(component_manager);
	});

	return new NodeFactory(entity_manager, component_manager);
});




var ComponentManager = (function(){
	function ComponentManager() {
		return this
	}

	ComponentManager.prototype.init = function(component_objects, components) {
		this.component_objects = component_objects;
		this.components = components;
	};

	ComponentManager.prototype.add_component_to_entity = function(entity_id, component_id, component_data){
		if(!this.component_objects[component_id]){
			this.component_objects[component_id] = {};
		}

		if(this.component_objects[component_id][entity_id]) {
			throw "Component with entity_id "+entity_id+" already has component_id " + component_id;
		}

		this.component_objects[component_id][entity_id] = new this.components[component_id](entity_id, component_data);
	}

	ComponentManager.prototype.update_component_for_entity = function(entity_id, component_id, component_data){
		if(!this.component_objects[component_id]){
			this.component_objects[component_id] = {}
		}
		if(!this.component_objects[component_id][entity_id]){
			this.add_component_to_entity(entity_id, component_id, component_data);
			return
		}

		this.component_objects[component_id][entity_id] = new this.components[component_id](entity_id, component_data);
	}

	ComponentManager.prototype.remove_component_from_entity = function(entity_id, component_id) {
		if(!this.component_objects[component_id]){
			return;
		}
		delete this.component_objects[component_id][entity_id];
	}

	ComponentManager.prototype.remove_all_components_from_entity = function(entity_id) {
		let all_components = Object.keys(components);
		for(let j in all_components) {
			let comp = all_components[j];
			if(this.component_objects[comp] && this.component_objects[comp][entity_id]) {
				delete this.component_objects[comp][entity_id];
			}
		}
	}

	ComponentManager.prototype.entity_has_component = function(entity_id, component_id) {
		if(!this.component_objects[component_id]){
			return false
		}
		return !!this.component_objects[component_id][entity_id];
	}

	ComponentManager.prototype.get_entities_with_component = function(component_id) {
		if(this.component_objects[component_id]) {
			return Object.keys(this.component_objects[component_id]);
		}
		return [];
	}

	ComponentManager.prototype.get_entities_with_components = function(components_list, entity_ids) {
		let intersection = [];

		var entities0 = entity_ids ? entity_ids : this.get_entities_with_component(components_list[0]);

		var start = entity_ids ? 0 : 1;

		for(let i = start; i < components_list.length; i++) {
			let entities = this.get_entities_with_component(components_list[i]);
			for(let j = 0; j < entities.length; j++) {
				let entity = entities[j];
				if (entities0.indexOf(entity) > -1) {
					intersection.push(entity);
				} 
			}
			entities0 = intersection;
			intersection = [];
		}
		intersection = entities0;
		return intersection;
	}

	ComponentManager.prototype.get_component_data_for_entity = function(entity_id, component_id) {
		if(!this.component_objects[component_id]) {
			return null;
		}
		return this.component_objects[component_id][entity_id];
	}

	return new ComponentManager();
})();
